controller/controller_server/csi_controller_server.py

```diff
@@ -534,6 +539,88 @@ def GetCapacity(self, request, context):
         logger.info("finished GetCapacity")
         return csi_pb2.GetCapacityResponse()
 
+    def ControllerExpandVolume(self, request, context):
+        set_current_thread_name(request.volume_id)
+        logger.info("ControllerExpandVolume")
+        secrets = request.secrets
+
+        try:
+            utils.validate_expand_volume_request(request)
+
+            user, password, array_addresses = utils.get_array_connection_info_from_secret(secrets)
+
+            try:
+                array_type, volume_id = utils.get_volume_id_info(request.volume_id)
+            except ObjectIdError as ex:
+                logger.warning("volume id is invalid. error : {}".format(ex))
+                context.set_code(grpc.StatusCode.INVALID_ARGUMENT)
+                return csi_pb2.ControllerExpandVolumeResponse()
+
+            with get_agent(user, password, array_addresses, array_type).get_mediator() as array_mediator:
+                logger.debug(array_mediator)
+
+                required_bytes = request.capacity_range.required_bytes
+                min_size = array_mediator.minimal_volume_size_in_bytes
+                max_size = array_mediator.maximal_volume_size_in_bytes
+
+                volume_before_expand = array_mediator.get_object_by_id(volume_id, config.VOLUME_TYPE_NAME)
+                if not volume_before_expand:
+                    raise controller_errors.ObjectNotFoundError(volume_id)
+
+                if required_bytes == 0 or volume_before_expand.capacity_bytes >= required_bytes:
```

Comment:
the `required_bytes == 0` check could theoretically save us an unnecessary call to `expand_volume`, e.g. if the sidecar would actually send 0 **and** if a negative capacity would return from a mediator (i.e. mediator bug).
but I'm not sure if this reason is good enough to keep this seemingly unnecessary check.
what do you think?

Reply:
the idea here is that if  `required_bytes == 0` then we dont care about the volume size.. but if you think that its redundant I can remove that check.

Reply:
I agree that 0 means that the volume size doesn't matter, so the code is more robust with this check, but we do call `get_object_by_id` (for `NOT_FOUND` error) anyway, so it doesn't add any other value.
and IMO, in this case, simplicity is more important than the insignificant additional "robustness".

Reply:
ok, done